shapeworks_cloud/core/rest.py

```diff
@@ -302,6 +304,105 @@ def download(self, request, **kwargs):
+        form_data = request.data
+        landmarks_info = form_data.get('info')
+        landmarks_locations = form_data.get('locations')
+
+        project.landmarks_info = landmarks_info
+        project_file_contents = json.loads(project.file.read())
+        project_file_contents['landmarks'] = landmarks_info
+        project.file.save(
+            project.file.name.split('/')[-1],
+            ContentFile(json.dumps(project_file_contents).encode()),
+        )
+        project.save()
+
+        ids_existing_with_coords = []
+        for subject_id, data in landmarks_locations.items():
+            target_subject = models.Subject.objects.get(id=subject_id)
+            for anatomy_type, locations in data.items():
+                landmarks_object, created = models.Landmarks.objects.get_or_create(
+                    project=project, subject=target_subject, anatomy_type=anatomy_type
+                )
+                file_content = ''
+                if (
+                    locations is not None
+                    and len(locations) > 0
+                    and locations[0] is not None
+                    and len(locations[0]) == 3
+                ):
+                    file_content = '\n'.join(
+                        ' '.join(str(n) for n in (loc.values() if isinstance(loc, dict) else loc))
+                        for loc in locations
+                    )
+                file_name = 'landmarks.csv'
+                if landmarks_object.file:
+                    file_name = landmarks_object.file.name.split('/')[-1]
+                landmarks_object.file.save(
+                    file_name,
+                    ContentFile(file_content.encode()),
+                )
+                ids_existing_with_coords.append(landmarks_object.id)
+
+        models.Landmarks.objects.filter(project=project).exclude(
+            id__in=ids_existing_with_coords
+        ).delete()
+
+        log_write_access(
+            timezone.now(),
+            self.request.user.username,
+            'Set Project Landmarks',
+            project.id,
+        )
+        return Response(serializers.ProjectReadSerializer(project).data)
+
+    @action(
+        detail=True,
+        url_path='constraints',
+        url_name='constraints',
+        methods=['POST'],
+    )
+    def set_constraints(self, request, **kwargs):
+        project = self.get_object()
+        form_data = request.data
+        constraints_locations = form_data.get('locations')
+
+        ids_existing_with_coords = []
+        for subject_id, data in constraints_locations.items():
+            target_subject = models.Subject.objects.get(id=subject_id)
+            for anatomy_type, locations in data.items():
+                constraints_object, created = models.Constraints.objects.get_or_create(
+                    project=project, subject=target_subject, anatomy_type=anatomy_type
+                )
+                file_name = 'constraints.json'
```

Comment:
Is there a reason constraints is saved as json and landmarks as csv?

Reply:
This is how they are stored with Studio, and I maintained that design to ensure compatibility.